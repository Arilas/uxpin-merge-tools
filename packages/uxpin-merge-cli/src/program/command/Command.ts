export enum Command {
  DUMP = 'dump',
  EXPERIMENT = 'experiment',
  INIT = 'init',
  CREATE_APP = 'create-app',
  GENERATE_PRESETS = 'generate-presets',
  PUSH = 'push',
  SERVER = 'server',
  SUMMARY = 'summary',
}

export const DEFAULT_COMMAND:Command = Command.EXPERIMENT;
