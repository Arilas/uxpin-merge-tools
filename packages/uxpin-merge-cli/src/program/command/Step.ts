import { DSMetadata } from '../DSMeta';

export interface Step {
  shouldRun:boolean;
  exec:(infos:DSMetadata) => any;
}
