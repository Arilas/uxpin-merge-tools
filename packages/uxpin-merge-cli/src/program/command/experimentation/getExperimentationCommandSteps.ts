import { Store } from '../../../program/utils/store/Store';
import { ExperimentProgramArgs } from '../../args/ProgramArgs';
import { Step } from '../Step';
import { experimentationBuildLibraryStep } from './steps/experimentationBuildLibraryStep';
import { experimentationCreateEpidStep } from './steps/experimentationCreateEpidStep';
import { experimentationOpenBrowserStep } from './steps/experimentationOpenBrowserStep';
import { experimentationPrepareMetadataStep } from './steps/experimentationPrepareMetadataStep';
import { experimentationRunServerStep } from './steps/experimentationRunServerStep';
import { experimentationRunNgrok } from './steps/experimentationRunNgrok';

export interface ExperimentationState {
  ngrokUrl:string|null,
}

const defaultState:ExperimentationState = {
  ngrokUrl: null,
};

export function getExperimentationCommandSteps(args:ExperimentProgramArgs):Step[] {
  const store:Store<ExperimentationState> = new Store(defaultState);

  return [
    experimentationBuildLibraryStep(args, store),
    experimentationCreateEpidStep(args, store),
    experimentationPrepareMetadataStep(args, store),
    experimentationRunNgrok(args, store),
    experimentationRunServerStep(args, store),
    experimentationOpenBrowserStep(args, store),
  ];
}
