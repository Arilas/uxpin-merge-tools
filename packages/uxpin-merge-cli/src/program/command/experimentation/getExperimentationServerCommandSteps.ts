import { startExperimentationServer } from '../../../steps/experimentation/server/startExperimentationServer';
import { ExperimentProgramArgs } from '../../args/ProgramArgs';
import { getTempDirPath } from '../../args/providers/paths/getTempDirPath';
import { DSMetadata } from '../../DSMeta';
import { Step } from '../Step';

export function getExperimentationServerCommandSteps(args:ExperimentProgramArgs):Step[] {
  return [
    { exec: thunkStartExperimentationServer(args), shouldRun: true },
  ];
}

function thunkStartExperimentationServer(args:ExperimentProgramArgs):(ds:DSMetadata) => Promise<any> {
  return () => {
    const { port, uxpinDomain } = args;
    return startExperimentationServer({
      port,
      uxpinDirPath: getTempDirPath(args),
      uxpinDomain,
    });
  };
}
