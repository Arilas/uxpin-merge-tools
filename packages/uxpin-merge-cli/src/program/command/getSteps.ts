import { BuildOptions } from '../../steps/building/BuildOptions';
import { ProgramArgs, PushProgramArgs, ServerProgramArgs } from '../ProgramArgs';
import { getDumpCommandSteps } from './dump/getDumpCommandSteps';
import { getPushCommandSteps } from './push/getPushCommandSteps';
import { getServerCommandSteps } from './server/getServerCommandSteps';
import { Step } from './Step';
import { getSummaryCommandSteps } from './summary/getSummaryCommandSteps';

export function getSteps(args:ProgramArgs):Step[] {
  if (args.command === 'server') {
    return getServerCommandSteps(getBuildOptions(args), args);
  }

  if (args.command === 'push') {
    return getPushCommandSteps(getBuildOptions(args));
  }

  if (args.command === 'dump') {
    return getDumpCommandSteps();
  }

  if (args.command === 'summary') {
    return getSummaryCommandSteps();
  }

  return [];
}

function getBuildOptions(args:ServerProgramArgs | PushProgramArgs):BuildOptions {
  const { cwd, webpackConfig, wrapper } = args;
  return {
    projectRoot: cwd,
    webpackConfigPath: webpackConfig,
    wrapperPath: wrapper,
  };
}
