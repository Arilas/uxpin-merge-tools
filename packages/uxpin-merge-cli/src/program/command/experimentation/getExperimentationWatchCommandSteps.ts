import { print } from '../../../utils/console/print';
import { printLine } from '../../../utils/console/printLine';
import { PrintColor } from '../../../utils/console/PrintOptions';
import { ExperimentProgramArgs } from '../../args/ProgramArgs';
import { Step } from '../Step';
import { experimentationBuildLibraryStep } from './steps/experimentationBuildLibraryStep';
import { experimentationPrepareMetadataStep } from './steps/experimentationPrepareMetadataStep';
import { ExperimentationState } from './getExperimentationCommandSteps';
import { Store } from '../../../program/utils/store/Store';

export const COMPILATION_SUCCESS_MESSAGE:string = 'üåê Compiled successfully! Now you can refresh your browser.';

const defaultState:ExperimentationState = {
  ngrokUrl: null,
};

export function getExperimentationWatchCommandSteps(args:ExperimentProgramArgs):Step[] {
  const store:Store<ExperimentationState> = new Store(defaultState);

  return [
    onCompilationStart(),
    experimentationBuildLibraryStep(args, store),
    experimentationPrepareMetadataStep(args, store),
    onCompilationEnd(),
  ];
}

function onCompilationStart():Step {
  return {
    exec: () => print('üé© Hang on to your hat! Compilation in progress...', { color: PrintColor.YELLOW }),
    shouldRun: true,
  };
}

function onCompilationEnd():Step {
  return {
    exec: () => printLine(`\r${COMPILATION_SUCCESS_MESSAGE}`),
    shouldRun: true,
  };
}
