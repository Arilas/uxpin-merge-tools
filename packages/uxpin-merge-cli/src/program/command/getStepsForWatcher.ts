import { ProgramArgs } from '../args/ProgramArgs';
import { getExperimentationWatchCommandSteps } from './experimentation/getExperimentationWatchCommandSteps';
import { Step } from './Step';

export function getStepsForWatcher(args:ProgramArgs):Step[] {
  if (args.command === 'experiment') {
    return getExperimentationWatchCommandSteps(args);
  }

  return [];
}
