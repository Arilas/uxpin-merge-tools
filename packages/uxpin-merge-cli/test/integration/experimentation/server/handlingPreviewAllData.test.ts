import { OK } from 'http-status-codes';
import { Response } from 'request';
import { setTimeoutBeforeAll } from '../../../utils/command/setTimeoutBeforeAll';
import { setupExperimentationServerTest } from '../../../utils/experimentation/setupExperimentationServerTest';

const CURRENT_TIMEOUT:number = 300000;
setTimeoutBeforeAll(CURRENT_TIMEOUT);

describe('Experimentation mode - handling preview all data', () => {
  const port:number = 8888;
  let response:Response;
  const { request } = setupExperimentationServerTest({
    port,
    projectPath: 'resources/designSystems/twoComponentsWithConfig',
    serverCmdArgs: [
      '--webpack-config "./webpack.config.js"',
      '--uxpin-api-domain "0.0.0.0:7448"',
    ],
  });

  beforeAll(async () => {
    const origin:string = 'https://app.uxpin.com';
    response = await request('/preview/all', { resolveWithFullResponse: true, headers: { origin } });
  });

  it('should responds with OK status code', async () => {
    expect(response.statusCode).toEqual(OK);
  });

  it('should responds with correct CORS headers and no-cache', async () => {
    // given
    const expectedHeaders:any = {
      'access-control-allow-credentials': 'true',
      'access-control-allow-headers': 'Origin, X-Requested-With, Content-Type, Accept, Range',
      'access-control-allow-origin': 'https://app.uxpin.com',
      'cache-control': 'no-cache',
      'content-type': 'application/json',
    };

    // then
    expect(response.headers).toEqual(expect.objectContaining(expectedHeaders));
  });

  it('should respond with proper body', () => {
    expect(JSON.parse(response.body)).toMatchSnapshot();
  });
});
