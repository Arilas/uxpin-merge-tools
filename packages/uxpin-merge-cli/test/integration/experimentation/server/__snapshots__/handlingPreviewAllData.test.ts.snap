// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Experimentation mode - handling preview all data should respond with proper body 1`] = `
Object {
  "breakpoints": Array [
    Object {
      "height": 768,
      "name": "Wide 1440",
      "version": 0,
      "width": 1440,
    },
    Object {
      "height": 480,
      "name": "iPhone (320px)",
      "version": 1,
      "width": 320,
    },
    Object {
      "height": 320,
      "name": "Phone Landscape (480px)",
      "version": 2,
      "width": 480,
    },
    Object {
      "height": 768,
      "name": "Tablets (600px)",
      "version": 3,
      "width": 600,
    },
    Object {
      "height": 1024,
      "name": "iPads (768px)",
      "version": 4,
      "width": 768,
    },
    Object {
      "height": 768,
      "name": "Standard websites (992px)",
      "version": 5,
      "width": 992,
    },
    Object {
      "height": 768,
      "name": "iPad Landscape (1024px)",
      "version": 6,
      "width": 1024,
    },
    Object {
      "height": 768,
      "name": "Wide websites (1224px)",
      "version": 7,
      "width": 1224,
    },
  ],
  "pageId": 1,
  "pages": Array [
    Object {
      "documentationExists": false,
      "id_page": 1,
      "is_active": true,
      "main_version": 0,
      "name": "Experiment",
      "parent": null,
      "sort_order": 0,
      "version_of": null,
      "version_type": 0,
    },
  ],
  "pagesData": Object {
    "1": Object {
      "canvasData": Object {
        "code_sync": Object {
          "bundles": Object {
            "3ab57996-fdf2-41cd-b3c6-85ba98596081_33a58bbfb9e97c671048f796c842723f13599762": "http://localhost:8888/code/library.js",
          },
          "components": Object {
            "ba14886c-2674-52a3-a147-7b88e725e4ee": Object {
              "componentId": "ba14886c-2674-52a3-a147-7b88e725e4ee",
              "info": Object {
                "dirPath": "src/components/Button",
                "implementation": Object {
                  "framework": "reactjs",
                  "lang": "typescript",
                  "path": "src/components/Button/Button.tsx",
                },
                "presets": Array [
                  Object {
                    "path": "src/components/Button/presets/0-default.json",
                  },
                ],
              },
              "name": "Button",
              "properties": Array [
                Object {
                  "description": "",
                  "isRequired": true,
                  "name": "children",
                  "type": Object {
                    "name": "node",
                    "structure": Object {},
                  },
                },
                Object {
                  "description": "",
                  "isRequired": false,
                  "name": "isPrimary",
                  "type": Object {
                    "name": "boolean",
                    "structure": Object {},
                  },
                },
              ],
              "revisionId": "3ab57996-fdf2-41cd-b3c6-85ba98596081_33a58bbfb9e97c671048f796c842723f13599762",
            },
            "f2dff102-3d25-5174-b733-12c4e58fdd5d": Object {
              "componentId": "f2dff102-3d25-5174-b733-12c4e58fdd5d",
              "info": Object {
                "dirPath": "src/components/Avatar",
                "implementation": Object {
                  "framework": "reactjs",
                  "lang": "typescript",
                  "path": "src/components/Avatar/Avatar.tsx",
                },
                "presets": Array [
                  Object {
                    "path": "src/components/Avatar/presets/0-default.json",
                  },
                ],
              },
              "name": "Avatar",
              "properties": Array [
                Object {
                  "description": "",
                  "isRequired": false,
                  "name": "children",
                  "type": Object {
                    "name": "node",
                    "structure": Object {},
                  },
                },
                Object {
                  "description": "",
                  "isRequired": true,
                  "name": "size",
                  "type": Object {
                    "name": "string",
                    "structure": Object {},
                  },
                },
                Object {
                  "description": "",
                  "isRequired": true,
                  "name": "imageUrl",
                  "type": Object {
                    "name": "string",
                    "structure": Object {},
                  },
                },
              ],
              "revisionId": "3ab57996-fdf2-41cd-b3c6-85ba98596081_33a58bbfb9e97c671048f796c842723f13599762",
            },
          },
          "presets": Object {
            "a7b05955-128e-5570-9968-26fa6d8f8705": Object {
              "componentRevisionId": "3ab57996-fdf2-41cd-b3c6-85ba98596081_33a58bbfb9e97c671048f796c842723f13599762_f2dff102-3d25-5174-b733-12c4e58fdd5d",
              "elements": Object {
                "1": Object {
                  "name": "Avatar",
                  "props": Object {
                    "imageUrl": "https://picsum.photos/100/100?image=1027",
                    "size": "small",
                  },
                },
              },
              "name": "default",
              "presetId": "a7b05955-128e-5570-9968-26fa6d8f8705",
              "rootId": "1",
              "sortIndex": 0,
            },
            "c383bd8a-caf4-5f90-a6b1-3471c355fc43": Object {
              "componentRevisionId": "3ab57996-fdf2-41cd-b3c6-85ba98596081_33a58bbfb9e97c671048f796c842723f13599762_ba14886c-2674-52a3-a147-7b88e725e4ee",
              "elements": Object {
                "1": Object {
                  "name": "Button",
                  "props": Object {
                    "children": "Click me",
                    "isPrimary": true,
                  },
                },
              },
              "name": "default",
              "presetId": "c383bd8a-caf4-5f90-a6b1-3471c355fc43",
              "rootId": "1",
              "sortIndex": 0,
            },
          },
        },
        "component_version": null,
        "components_master_ids": Array [],
        "components_versions": Array [],
        "components_versions_map": Array [],
        "is_component": "0",
        "last_update": "0",
        "page": Object {
          "canvas": Object {
            "props": Object {
              "storedElements": Array [],
            },
            "type": "Canvas",
            "v": "2.0",
          },
        },
      },
      "metaData": Object {
        "content": "/api/projects/3ab57996-fdf2-41cd-b3c6-85ba98596081_33a58bbfb9e97c671048f796c842723f13599762/pages/1/content",
        "size": Object {
          "height": 768,
          "scrollx": false,
          "scrolly": false,
          "width": 1440,
        },
      },
    },
  },
  "redirect": false,
}
`;
